File: search-insert-position/search-insert-position.java
class Solution {
    public int searchInsert(int[] nums, int target) {
       int start = 0;
       int end = nums.length - 1;
       int possible = 0;
       while (start <= end) {
           int middle = start + (end - start)/2;
           if (nums[middle] == target) {
               return middle;
           } else if (nums[middle] < target) {
               start = middle + 1;
               int diff2 = target - nums[possible];
               if (diff2 < 0) {
                   diff2 = -diff2;
               }
               if (diff2 >= target - nums[middle] ) {
                   possible = middle + 1;
               }
           } else if (nums[middle] > target) {
               end = middle - 1;
               int diff2 = nums[possible] - target;
               if (diff2 < 0) {
                   diff2 = -diff2;
               }
               if (diff2 >= nums[middle] - target) {
                   possible = middle;
               }               
           }
           System.out.println("start " + start);
           System.out.println("end " + end);
           System.out.println("middle " + middle);
           System.out.println("possible " + possible);
           
       }
       return possible; 
    }
}

File: 35-search-insert-position/search-insert-position.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        possible = 0
        higher = len(nums) - 1
        lower = 0
        while lower <= higher:
            middle = (higher + lower) // 2
            if (nums[middle] == target):
                return middle
            elif (nums[middle] > target):
                if (nums[middle - 1] < target):
                    possible = middle
                else:
                    possible = middle - 1
                    if possible < 0:
                        possible = 0
                higher = middle - 1
            elif (nums[middle] < target):
                possible = middle + 1
                lower = middle + 1
        return possible